File: frameworks/quarkus-graphql/src/main/java/com/velocitybench/repository/PostRepository.java
package com.velocitybench.repository;

import com.velocitybench.model.Post;
import io.agroal.api.AgroalDataSource;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.sql.*;
import java.util.*;

@ApplicationScoped
public class PostRepository {

    @Inject
    AgroalDataSource dataSource;

    public Optional<Post> findById(String id) {
        String sql = "SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post WHERE id = ?";
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setObject(1, UUID.fromString(id));
            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return Optional.of(mapRow(rs));
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return Optional.empty();
    }

    public Map<Integer, Post> findByPks(Set<Integer> pks) {
        if (pks.isEmpty()) return Collections.emptyMap();

        String placeholders = String.join(",", Collections.nCopies(pks.size(), "?"));
        String sql = "SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post WHERE pk_post IN (" + placeholders + ")";

        Map<Integer, Post> result = new HashMap<>();
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            int i = 1;
            for (Integer pk : pks) {
                stmt.setInt(i++, pk);
            }
            try (ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    Post post = mapRow(rs);
                    result.put(post.getPkPost(), post);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return result;
    }

    public Map<Integer, List<Post>> findByAuthorPks(Set<Integer> authorPks, int limit) {
        if (authorPks.isEmpty()) return Collections.emptyMap();

        String placeholders = String.join(",", Collections.nCopies(authorPks.size(), "?"));
        String sql = """
            SELECT * FROM (
                SELECT pk_post, id, fk_author, title, content, created_at, updated_at,
                       ROW_NUMBER() OVER (PARTITION BY fk_author ORDER BY pk_post) as rn
                FROM tb_post WHERE fk_author IN (%s)
            ) t WHERE rn <= ?
            ORDER BY fk_author, pk_post
            """.formatted(placeholders);

        Map<Integer, List<Post>> result = new HashMap<>();
        for (Integer pk : authorPks) {
            result.put(pk, new ArrayList<>());
        }

        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            int i = 1;
            for (Integer pk : authorPks) {
                stmt.setInt(i++, pk);
            }
            stmt.setInt(i, limit);
            try (ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    Post post = mapRow(rs);
                    result.get(post.getFkAuthor()).add(post);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return result;
    }

    public List<Post> findAll(int limit) {
        String sql = "SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post ORDER BY pk_post LIMIT ?";
        List<Post> posts = new ArrayList<>();
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, Math.min(limit, 100));
            try (ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    posts.add(mapRow(rs));
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return posts;
    }

    public Optional<Post> update(String id, String title, String content) {
        StringBuilder sql = new StringBuilder("UPDATE tb_post SET updated_at = NOW()");
        List<Object> params = new ArrayList<>();

        if (title != null) {
            sql.append(", title = ?");
            params.add(title);
        }
        if (content != null) {
            sql.append(", content = ?");
            params.add(content);
        }

        sql.append(" WHERE id = ?");
        params.add(UUID.fromString(id));

        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) {
                stmt.setObject(i + 1, params.get(i));
            }
            stmt.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return findById(id);
    }

    private Post mapRow(ResultSet rs) throws SQLException {
        return new Post(
            rs.getInt("pk_post"),
            rs.getObject("id", UUID.class),
            rs.getInt("fk_author"),
            rs.getString("title"),
            rs.getString("content"),
            rs.getTimestamp("created_at").toLocalDateTime(),
            rs.getTimestamp("updated_at") != null ? rs.getTimestamp("updated_at").toLocalDateTime() : null
        );
    }
}

File: frameworks/play-graphql/app/repositories/PostRepository.scala
package repositories

import db.Database
import models.Post
import javax.inject.{Inject, Singleton}
import java.util.UUID
import scala.collection.mutable

@Singleton
class PostRepository @Inject()(database: Database) {

  def findById(id: String): Option[Post] = {
    val conn = database.getConnection
    try {
      val stmt = conn.prepareStatement(
        "SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post WHERE id = ?"
      )
      stmt.setObject(1, UUID.fromString(id))
      val rs = stmt.executeQuery()
      if (rs.next()) Some(mapRow(rs)) else None
    } finally {
      conn.close()
    }
  }

  def findByPks(pks: Set[Int]): Map[Int, Post] = {
    if (pks.isEmpty) return Map.empty

    val conn = database.getConnection
    try {
      val placeholders = pks.map(_ => "?").mkString(",")
      val stmt = conn.prepareStatement(
        s"SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post WHERE pk_post IN ($placeholders)"
      )
      pks.zipWithIndex.foreach { case (pk, idx) => stmt.setInt(idx + 1, pk) }
      val rs = stmt.executeQuery()
      val result = mutable.Map[Int, Post]()
      while (rs.next()) {
        val post = mapRow(rs)
        result(post.pkPost) = post
      }
      result.toMap
    } finally {
      conn.close()
    }
  }

  def findByAuthorPks(authorPks: Set[Int], limit: Int): Map[Int, Seq[Post]] = {
    if (authorPks.isEmpty) return Map.empty

    val conn = database.getConnection
    try {
      val placeholders = authorPks.map(_ => "?").mkString(",")
      val sql =
        s"""SELECT * FROM (
           |  SELECT pk_post, id, fk_author, title, content, created_at, updated_at,
           |         ROW_NUMBER() OVER (PARTITION BY fk_author ORDER BY pk_post) as rn
           |  FROM tb_post WHERE fk_author IN ($placeholders)
           |) t WHERE rn <= ?
           |ORDER BY fk_author, pk_post""".stripMargin

      val stmt = conn.prepareStatement(sql)
      authorPks.zipWithIndex.foreach { case (pk, idx) => stmt.setInt(idx + 1, pk) }
      stmt.setInt(authorPks.size + 1, limit)

      val rs = stmt.executeQuery()
      val result = mutable.Map[Int, mutable.Buffer[Post]]()
      authorPks.foreach(pk => result(pk) = mutable.Buffer[Post]())

      while (rs.next()) {
        val post = mapRow(rs)
        result(post.fkAuthor) += post
      }
      result.view.mapValues(_.toSeq).toMap
    } finally {
      conn.close()
    }
  }

  def findAll(limit: Int): Seq[Post] = {
    val conn = database.getConnection
    try {
      val stmt = conn.prepareStatement(
        "SELECT pk_post, id, fk_author, title, content, created_at, updated_at FROM tb_post ORDER BY pk_post LIMIT ?"
      )
      stmt.setInt(1, Math.min(limit, 100))
      val rs = stmt.executeQuery()
      val posts = mutable.Buffer[Post]()
      while (rs.next()) {
        posts += mapRow(rs)
      }
      posts.toSeq
    } finally {
      conn.close()
    }
  }

  def update(id: String, title: Option[String], content: Option[String]): Option[Post] = {
    val conn = database.getConnection
    try {
      val sets = mutable.Buffer("updated_at = NOW()")
      val params = mutable.Buffer[Any]()

      title.foreach { t =>
        sets += "title = ?"
        params += t
      }
      content.foreach { c =>
        sets += "content = ?"
        params += c
      }

      val sql = s"UPDATE tb_post SET ${sets.mkString(", ")} WHERE id = ?"
      val stmt = conn.prepareStatement(sql)
      params.zipWithIndex.foreach { case (p, idx) => stmt.setObject(idx + 1, p) }
      stmt.setObject(params.size + 1, UUID.fromString(id))
      stmt.executeUpdate()
    } finally {
      conn.close()
    }
    findById(id)
  }

  private def mapRow(rs: java.sql.ResultSet): Post = {
    Post(
      pkPost = rs.getInt("pk_post"),
      id = java.util.UUID.fromString(rs.getString("id")),
      fkAuthor = rs.getInt("fk_author"),
      title = rs.getString("title"),
      content = Option(rs.getString("content")),
      createdAt = rs.getTimestamp("created_at").toLocalDateTime,
      updatedAt = Option(rs.getTimestamp("updated_at")).map(_.toLocalDateTime)
    )
  }
}
